axelor-business-project/src/main/java/com/axelor/apps/businessproject/listener/ProjectTaskListener.java

```diff
@@ -0,0 +1,43 @@
+package com.axelor.apps.businessproject.listener;
+
+import com.axelor.apps.project.db.ProjectTask;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.persistence.PreUpdate;
+
+public class ProjectTaskListener {
+  @PreUpdate
+  public void onProjectTaskPreUpdate(ProjectTask projectTask) {
+    calculateParentTaskProgress(projectTask);
+  }
+
+  public void calculateParentTaskProgress(ProjectTask projectTask) {
+    ProjectTask parentTask = projectTask.getParentTask();
+    if (parentTask != null) {
+      List<ProjectTask> childProjectTasks = parentTask.getProjectTaskList();
+      childProjectTasks =
+          childProjectTasks.stream()
+              .map(task -> Objects.equals(task.getId(), projectTask.getId()) ? projectTask : task)
+              .collect(Collectors.toList());
+
+      BigDecimal sumProgressTimesPlanifiedTime =
+          childProjectTasks.stream()
+              .map(task -> task.getProgress().multiply(task.getPlannedTime()))
+              .reduce(BigDecimal.ZERO, BigDecimal::add);
+
+      BigDecimal sumPlannedTime =
+          childProjectTasks.stream()
+              .map(ProjectTask::getPlannedTime)
+              .reduce(BigDecimal.ZERO, BigDecimal::add);
+      int scale = 2;
+      BigDecimal averageProgress =
+          sumPlannedTime.compareTo(BigDecimal.ZERO) != 0
+              ? sumProgressTimesPlanifiedTime.divide(sumPlannedTime, scale, RoundingMode.HALF_UP)
+              : BigDecimal.ZERO;
+      parentTask.setProgress(averageProgress);
+    }
+  }
+}
```

Comment:
This should be also called in the repository of the ProjectTask instead of using a Listener. Using a listener could cause some performance issues and would be called when updating a ProjectTask even if the progress is not updated.